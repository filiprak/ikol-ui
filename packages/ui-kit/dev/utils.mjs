import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
