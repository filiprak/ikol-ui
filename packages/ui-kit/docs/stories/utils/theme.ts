import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
