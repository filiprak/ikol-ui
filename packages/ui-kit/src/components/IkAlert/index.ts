import IkAlert from './IkAlert.vue';
export interface IkAlert extends InstanceType<typeof IkAlert> { };
export { IkAlert };
