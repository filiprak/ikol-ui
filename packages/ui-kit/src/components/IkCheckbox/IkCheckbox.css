.ik-checkbox__label {
    min-height: 20px;
    display: flex;
    cursor: pointer;
    margin-bottom: 5px;
}

.ik-checkbox--disabled {
    pointer-events: none;
    cursor: not-allowed;
    opacity: 0.55;
}

.ik-checkbox__label input {
    visibility: hidden;
    display: none;
}

.ik-checkbox__label input:checked + .ik-checbox__icon {
    border-color: var(--primary-color);
    color: var(--primary-color);
}

.ik-checbox__icon {
    color: rgba(143, 155, 179, 0.5);
    font-size: 13px;
    margin-right: 5px;
    height: 20px;
    width: 20px;
    cursor: pointer;
    vertical-align: middle;
    border: 2px solid rgba(143, 155, 179, 0.5);
    border-radius: 3px;
    text-align: center;
    flex-shrink: 0;
    flex-grow: 0;
}

.ik-checkbox__text {
    font-weight: normal;
}

/* basic variant styles */
.ik-checkbox--basic .ik-checkbox__label input:checked + .ik-checbox__icon {
    border-color: var(--primary-color);
    color: var(--primary-color);
}

/* switch variant styles */
.ik-checkbox--switch {
    line-height: 28px;
}

.ik-checkbox__switch {
    position: relative;
    flex: 0 0 auto;
    height: 28px;
    width: 56px;
    border-radius: 56px;
    padding: 4px;
    cursor: pointer;
    background-color: rgba(143, 155, 179, 0.3);

    transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
    -webkit-transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
    -moz-transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
    -o-transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
}

.ik-checkbox__text:not(:empty) {
    margin-left: 10px;
}

.ik-checkbox--switch .ik-checkbox__label {
    margin: 0;
    display: inline-flex;
}

.ik-checkbox--switch .ik-checkbox__label input:checked + .ik-checkbox__switch {
    background-color: var(--primary-color);
}

.ik-checkbox--switch .ik-checkbox__label input:checked + .ik-checkbox__switch > div {
    left: calc(100% - 24px);
}

.ik-checkbox__switch > div {
    top: 4px;
    left: 4px;
    position: absolute;
    background-color: var(--on-primary-color);
    width: 20px;
    height: 20px;
    border-radius: 20px;

    transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
    -webkit-transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
    -moz-transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
    -o-transition: 0.2s cubic-bezier(0.785, 0.135, 0.150, 0.860);
}

.ik-checkbox--basic .ik-checkbox__text,
.ik-checkbox--switch .ik-checkbox__text {
    font-weight: 600;
}

.ik-checkbox--no-margin .ik-checkbox__label {
    margin-bottom: 0;
}
