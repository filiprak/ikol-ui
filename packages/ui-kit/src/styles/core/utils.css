/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
