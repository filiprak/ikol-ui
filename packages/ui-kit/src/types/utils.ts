import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
