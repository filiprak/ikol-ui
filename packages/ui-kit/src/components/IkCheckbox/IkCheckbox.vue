<template>
    <div :class="root_class">
        <label class="ik-checkbox__label">
            <input v-model="internal_checked"
                   type="checkbox"
                   :name="name"
                   :checked="internal_checked" />
            <template v-if="variant === 'switch'">
                <div class="ik-checkbox__switch">
                    <div></div>
                </div>
            </template>
            <template v-else>
                <div class="ik-checbox__icon">
                    <IkIcon v-if="internal_checked || props.indeterminate"
                            :icon="$props.indeterminate ? 'minus' : 'check'"
                            :title="title" />
                </div>
            </template>
            <div class="ik-checkbox__text">
                <slot></slot>
            </div>
        </label>
    </div>
</template>
<script setup lang="ts">
import '@ui/styles';
import './IkCheckbox.css';
import { computed, type PropType } from 'vue';
import { IkIcon } from '@ui/components/IkIcon';
import { isArray, isBoolean, isString } from '@ui/utils/helpers';

const props = defineProps({
    /**
     * Current state of checkbox. This is also used for `v-model` binding.
     * Possible types are `"array", "boolean" , "string"`.
     * - `"array"` option allows to connect multiple checkboxes to single `v-model` variable
     * - `"string"` option allows to connect multiple exclusive checkboxes to single `v-model` variable
     * - `"boolean"` option allows to connect checkbox to simple `true/false` `v-model` variable
     * See examples below.
     **/
    modelValue: {
        type: [String, Boolean, Array] as PropType<boolean | string | string[]>,
        default: false,
    },
    /**
    * Sets `name` attribute for native html input element. Is also as id for 'array and 'string' v-model types.
    **/
    name: {
        type: String,
    },
    /**
     * Tooltip text shown when mouse hovers checkbox icon.
    **/
    title: {
        type: String,
    },
    disabled: {
        type: Boolean,
        default: false,
    },
    /**
     * Renders checkbox with `minus` icon. This applies only to `"basic"` variant.
     * Indeterminate state can be used to present checkbox with unknown state.
    **/
    indeterminate: {
        type: Boolean,
        default: false,
    },
    /**
     * Style variant. Possible values `"basic", "switch"`.
     * - `"basic"` is regular checkbox with check icon
     * - `"switch"` is modern styled switch
    **/
    variant: {
        type: String as PropType<'basic' | 'switch'>,
        default: 'basic',
    },
    no_margin: {
        type: Boolean,
        default: false,
    },
});

const emit = defineEmits<{
    (e: 'update:modelValue', value: boolean | string | string[]): void
}>();

const root_class = computed(() => {
    const classes: Record<string, boolean> = {
        'ik-checkbox': true,
        'ik-checkbox--disabled': props.disabled,
        'ik-checkbox--no-margin': props.no_margin,
    };
    if (props.variant) {
        classes['ik-checkbox--' + props.variant] = true;
    }
    return classes;
});

const internal_checked = computed({
    get: () => {
        if (isString(props.modelValue)) {
            return props.modelValue === props.name;
        } else if (isArray(props.modelValue)) {
            if (!props.name) {
                return false;
            }
            return props.modelValue.includes(props.name);
        } else if (isBoolean(props.modelValue)) {
            return props.modelValue;
        } else {
            return false;
        }
    },
    set: (val) => {
        if (isString(props.modelValue)) {
            emit('update:modelValue', val && props.name ? props.name : '');
        } else if (isArray(props.modelValue)) {
            if (val) {
                if (!props.name) {
                    return emit('update:modelValue', props.modelValue);
                }
                if (!props.modelValue.includes(props.name)) {
                    emit('update:modelValue', [...props.modelValue, props.name]);
                }
            } else {
                const arr = props.modelValue.filter(name => name !== props.name);
                emit('update:modelValue', arr);
            }
        } else {
            emit('update:modelValue', val);
        }
    },
});

</script>
