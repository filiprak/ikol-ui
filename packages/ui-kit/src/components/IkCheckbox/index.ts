import IkCheckbox from './IkCheckbox.vue';
export interface IkCheckbox extends InstanceType<typeof IkCheckbox> { };
export { IkCheckbox };
