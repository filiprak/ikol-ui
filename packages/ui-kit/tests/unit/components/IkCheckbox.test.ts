import { renderShallow, type RenderOptions } from '@ui/tests/tools/utils';
import { IkCheckbox } from '@ui/components/IkCheckbox';

describe('IkCheckbox', () => {
    const getInstance = (options: RenderOptions<typeof IkCheckbox> = {}) => renderShallow(IkCheckbox, {
        attachTo: document.body,
    }, options);

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toStrictEqual([
            'ik-checkbox',
            'ik-checkbox--basic',
        ]);
        expect(wrapper.vm.internal_checked).toBe(false);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: '<span>This is checkbox text...</span>',
            },
        });

        expect(wrapper.text()).toContain('This is checkbox text');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders checked', () => {
        const wrapper = getInstance({
            props: {
                modelValue: true,
            },
        });

        expect(wrapper.vm.internal_checked).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders disabled', () => {
        const wrapper = getInstance({
            props: {
                disabled: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-checkbox--disabled');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders indeterminate', () => {
        const wrapper = getInstance({
            props: {
                indeterminate: true,
            },
        });

        expect(wrapper.find('ik-icon-stub').attributes().icon).toBe('minus');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('emits event when toggled', () => {
        const wrapper = getInstance({
            props: {
                modelValue: true,
                name: 'checked1',
            },
        });

        wrapper.find('input').trigger('click');
        wrapper.find('input').trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [false],
            [true],
        ]);
    });

    it('emits event when string type', async () => {
        const wrapper = getInstance({
            props: {
                modelValue: 'string',
                name: 'test-1',
            },
        });

        wrapper.find('input').trigger('click');
        wrapper.find('input').trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            ['test-1'],
            [''],
        ]);
    });

    it('emits event when string type with initial model', () => {
        const wrapper = getInstance({
            props: {
                name: 'test-1',
                modelValue: 'test-1',
            },
        });

        wrapper.find('input').trigger('click');
        wrapper.find('input').trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [''],
            ['test-1'],
        ]);
    });

    it('emits event when string type with initial model 2', () => {
        const wrapper = getInstance({
            props: {
                name: 'test-1',
                modelValue: 'xyz',
            },
        });

        wrapper.find('input').trigger('click');
        wrapper.find('input').trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            ['test-1'],
            [''],
        ]);
    });

    it('emits event when array type', () => {
        const wrapper = getInstance({
            props: {
                name: 'test-1',
                modelValue: [],
            },
        });

        wrapper.find('input').trigger('click');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [['test-1']],
        ]);
    });

    it('emits event when array type with initial model', async () => {
        const wrapper = getInstance({
            props: {
                name: 'test-1',
                modelValue: ['a', 'b', 'test-1'],
            },
        });

        wrapper.find('input').trigger('click');
        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [['a', 'b']],
        ]);

        await wrapper.setProps({ modelValue: ['a', 'b'] });

        wrapper.find('input').trigger('click');
        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            [['a', 'b']],
            [['a', 'b', 'test-1']],
        ]);
    });

    it('renders in basic variant', () => {
        const wrapper = getInstance({
            props: {
                variant: 'basic',
            },
        });

        expect(wrapper.classes()).toContain('ik-checkbox--basic');
    });

    it('renders in switch variant', () => {
        const wrapper = getInstance({
            props: {
                variant: 'switch',
            },
        });

        expect(wrapper.classes()).toContain('ik-checkbox--switch');
        expect(wrapper.find('.ik-checkbox__switch').isVisible()).toBeTruthy();
    });

    it('renders without label margin', () => {
        const wrapper = getInstance({
            props: {
                no_margin: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-checkbox--no-margin');
    });
});
