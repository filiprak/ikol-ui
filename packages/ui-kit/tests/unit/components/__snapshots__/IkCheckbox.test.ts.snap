// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkCheckbox renders checked 1`] = `
<div
  class="ik-checkbox ik-checkbox--basic"
>
  <label
    class="ik-checkbox__label"
  >
    <input
      type="checkbox"
    />
    <div
      class="ik-checbox__icon"
    >
      <ik-icon-stub
        circle="false"
        disabled="false"
        icon="check"
        tag="i"
      />
    </div>
    <div
      class="ik-checkbox__text"
    >
      
      
    </div>
  </label>
</div>
`;

exports[`IkCheckbox renders disabled 1`] = `
<div
  class="ik-checkbox ik-checkbox--disabled ik-checkbox--basic"
>
  <label
    class="ik-checkbox__label"
  >
    <input
      type="checkbox"
    />
    <div
      class="ik-checbox__icon"
    >
      <!--v-if-->
    </div>
    <div
      class="ik-checkbox__text"
    >
      
      
    </div>
  </label>
</div>
`;

exports[`IkCheckbox renders indeterminate 1`] = `
<div
  class="ik-checkbox ik-checkbox--basic"
>
  <label
    class="ik-checkbox__label"
  >
    <input
      type="checkbox"
    />
    <div
      class="ik-checbox__icon"
    >
      <ik-icon-stub
        circle="false"
        disabled="false"
        icon="minus"
        tag="i"
      />
    </div>
    <div
      class="ik-checkbox__text"
    >
      
      
    </div>
  </label>
</div>
`;

exports[`IkCheckbox renders with correct default state 1`] = `
<div
  class="ik-checkbox ik-checkbox--basic"
>
  <label
    class="ik-checkbox__label"
  >
    <input
      type="checkbox"
    />
    <div
      class="ik-checbox__icon"
    >
      <!--v-if-->
    </div>
    <div
      class="ik-checkbox__text"
    >
      
      
    </div>
  </label>
</div>
`;

exports[`IkCheckbox renders with default slot 1`] = `
<div
  class="ik-checkbox ik-checkbox--basic"
>
  <label
    class="ik-checkbox__label"
  >
    <input
      type="checkbox"
    />
    <div
      class="ik-checbox__icon"
    >
      <!--v-if-->
    </div>
    <div
      class="ik-checkbox__text"
    >
      
      <span>
        This is checkbox text...
      </span>
      
    </div>
  </label>
</div>
`;
